File: Java/src/main/java/minimal/CookingService.java
package minimal;

import java.util.ArrayList;
import java.util.List;

import minimal.dao.Database;
import minimal.model.Cookbook;
import minimal.model.Ingredient;
import minimal.model.Name;
import minimal.model.NoSuchIngredientException;
import minimal.model.Recipe;

public class CookingService {

    private final Database database = new Database();

    public List<Name> listIngredients() {
        return namesOf(getIngredients());
    }

    private List<Ingredient> getIngredients() {
        return database.selectIngredients();
    }

    private List<Name> namesOf(List<Ingredient> ingredients) {
        List<Name> names = new ArrayList<>();
        for (Ingredient ingredient : ingredients) {
            names.add(ingredient.name());
        }
        return names;
    }

    public Ingredient takeOut(Name name) {
        Ingredient ingredient = findIngredientWith(name);
        ingredient.ensureOpened();
        return ingredient;
    }

    private Ingredient findIngredientWith(Name name) {
        for (Ingredient ingredient : getIngredients()) {
            if (ingredient.name().isNamedLike(name)) {
                return ingredient;
            }
        }
        throw new NoSuchIngredientException(name);
    }

    private List<Ingredient> takeOut(List<Name> listOfIngredients) {
        List<Ingredient> ingredients = new ArrayList<>();
        for (Name name : listOfIngredients) {
            ingredients.add(takeOut(name));
        }
        return ingredients;
    }

    public Cookbook getCookbook() {
        return new Cookbook(listAllRecipes());
    }

    private List<Name> listAllRecipes() {
        return database.selectAllRecipies();
    }

    public Recipe howToCook(Name meal) {
        return findRecipeFor(meal);
    }

    private Recipe findRecipeFor(Name meal) {
        return database.selectRecipeFor(meal);
    }

    public List<Ingredient> takeOutAllFor(Recipe recipe) {
        return takeOut(recipe.listIngredients());
    }

    public void addToMyCookbook(Recipe recipe) {
        database.insertRecipe(recipe);
    }

    public void removeFromCookbook(Name nameOfMeal) {
        database.removeRecipe(nameOfMeal);
    }
}

File: PHP/src/CookingService.php
<?php

namespace Minimal;

use Minimal\Dao\Database;
use Minimal\Model\Cookbook;
use Minimal\Model\Ingredient;
use Minimal\Model\Name;
use Minimal\Model\Recipe;
use Minimal\Model\WeCanNotCookAnythingException;

class CookingService
{
    private Database $database;

    public function __construct()
    {
        $this->database = new Database();
    }

    public function listIngredients(): array
    {
        return $this->namesOf($this->getIngredients());
    }

    private function getIngredients(): array
    {
        return $this->database->selectIngredients();
    }

    private function namesOf(array $ingredients): array
    {
        $names = [];
        foreach ($ingredients as $ingredient) {
            $names[] = $ingredient->name();
        }
        return $names;
    }

    public function takeOut(Name $name): Ingredient
    {
        $ingredient = $this->findIngredientWith($name);
        $ingredient->ensureOpened();
        return $ingredient;
    }

    private function findIngredientWith(Name $name): Ingredient
    {
        $ingredients = $this->getIngredients();
        foreach ($ingredients as $ingredient) {
            if ($ingredient->name()->isNamedLike($name)) {
                return $ingredient;
            }
        }
        throw new NoSuchIngredientException($name);
    }

    private function takeOutAll(array $listOfIngredients): array
    {
        $ingredients = [];
        foreach ($listOfIngredients as $name) {
            $ingredients[] = $this->takeOut($name);
        }
        return $ingredients;
    }

    public function getCookbook(): Cookbook
    {
        return new Cookbook($this->listAllRecipes());
    }

    private function listAllRecipes(): array
    {
        return $this->database->selectAllRecipes();
    }

    public function howToCook(Name $meal): Recipe
    {
        return $this->findRecipeFor($meal);
    }

    private function findRecipeFor(Name $meal): Recipe
    {
        return $this->database->selectRecipeFor($meal);
    }

    public function takeOutAllFor(Recipe $recipe): array
    {
        return $this->takeOutAll($recipe->listIngredients());
    }

    public function addToMyCookbook(Recipe $recipe)
    {
        $this->database->insertRecipe($recipe);
    }

    public function removeFromCookbook(Name $nameOfMeal)
    {
        $this->database->removeRecipe($nameOfMeal);
    }
}
